src/coreclr/interpreter/compiler.cpp

```diff
@@ -1289,56 +1426,111 @@ void InterpCompiler::CreateILVars()
         INTERP_DUMP("alloc local var %d to offset %d\n", index, offset);
         offset += size;
         sigArg = m_compHnd->getArgNext(sigArg);
+        index++;
     }
-    offset = ALIGN_UP_TO(offset, INTERP_STACK_ALIGNMENT);
 
+    offset = ALIGN_UP_TO(offset, INTERP_STACK_ALIGNMENT);
     m_ILLocalsSize = offset - m_ILLocalsOffset;
+
+    INTERP_DUMP("\nCreate clause Vars:\n");
+
+    m_clauseVarsIndex = index;
+
+    for (unsigned int i = 0; i < m_methodInfo->EHcount; i++)
+    {
+        new (&m_pVars[index]) InterpVar(InterpTypeO, NULL, INTERP_STACK_SLOT_SIZE);
+        m_pVars[index].global = true;
+        m_pVars[index].ILGlobal = true;
+        m_pVars[index].offset = offset;
+        INTERP_DUMP("alloc clause var %d to offset %d\n", index, offset);
+        offset += INTERP_STACK_SLOT_SIZE;
+        index++;
+    }
+
     m_totalVarsStackSize = offset;
 }
 
-bool InterpCompiler::CreateBasicBlocks(CORINFO_METHOD_INFO* methodInfo)
+void InterpCompiler::CreateFinallyCallIslandBasicBlocks(CORINFO_METHOD_INFO* methodInfo, int32_t leaveOffset, InterpBasicBlock* pLeaveTargetBB)
 {
-    int32_t codeSize = methodInfo->ILCodeSize;
-    uint8_t *codeStart = methodInfo->ILCode;
-    uint8_t *codeEnd = codeStart + codeSize;
-    const uint8_t *ip = codeStart;
-
-    m_ppOffsetToBB = (InterpBasicBlock**)AllocMemPool0(sizeof(InterpBasicBlock*) * (methodInfo->ILCodeSize + 1));
-    GetBB(0);
-
+    // Create finally call island basic blocks for all try regions with finally clauses that the leave exits.
+    // That means when the leaveOffset is inside the try region and the target is outside of it.
+    // These finally call island blocks are used for non-exceptional finally execution.
+    bool firstFinallyCallIsland = true;
+    InterpBasicBlock* pInnerFinallyCallIslandBB = NULL;
     for (unsigned int i = 0; i < methodInfo->EHcount; i++)
     {
         CORINFO_EH_CLAUSE clause;
         m_compHnd->getEHinfo(methodInfo->ftn, i, &clause);
+        if (clause.Flags != CORINFO_EH_CLAUSE_FINALLY)
+        {
+            continue;
+        }
 
-        if ((codeStart + clause.TryOffset) > codeEnd ||
-                (codeStart + clause.TryOffset + clause.TryLength) > codeEnd)
+        if ((uint32_t)leaveOffset < clause.TryOffset || (uint32_t)leaveOffset > (clause.TryOffset + clause.TryLength))
         {
-            return false;
+            continue;
         }
-        GetBB(clause.TryOffset);
 
-        if ((codeStart + clause.HandlerOffset) > codeEnd ||
-                (codeStart + clause.HandlerOffset + clause.HandlerLength) > codeEnd)
+        if ((uint32_t)pLeaveTargetBB->ilOffset >= clause.TryOffset && (uint32_t)pLeaveTargetBB->ilOffset <= (clause.TryOffset + clause.TryLength))
```

Comment:
clause.TryOffset + TryLength here - iirc this is a case where TryLength is the end offset, right?